### PR TITLE
Support flattening common options from a struct (#21)

### LINKED ISSUE #21: Flatten
Add ability to reuse struct fields via `flatten` attribute similar to [serde-rs/serde#119](https://github.com/serde-rs/serde/issues/119). Example: 
```rust
#[derive(StructOpt)]
#[structopt(name = "cargo", about = "the rust package manager")]
enum Cargo {
    #[structopt(name = "build")]
    Build {
        #[structopt(flatten)]
        common: CommonFlags,
        // some build-specific flags
    },
    #[structopt(name = "test")]
    Test {
        #[structopt(flatten)]
        common: CommonFlags,
        // some test-specific flags
    },
}
#[derive(StructOpt)]
struct CommonFlags {
    #[structopt(name = "release")]
    release: bool,
    // other common flags 
}
```

### PR DESCRIPTION
**Pull request by birkenfeld**
Fix #21

### CODE REVIEW

File: structopt-derive/src/lib.rs [hunk 1]
```diff
@@ -84,6 +84,14 @@ fn gen_augmentation(fields: &Punctuated<Field, Comma>, app_var: &Ident) -> quote
         .filter_map(|field| {
             let attrs = Attrs::from_field(field);
             if attrs.is_subcommand() { return None; }
+
+            if attrs.should_flatten() {
+                let ty = &field.ty;
+                return Some(quote!{
+                    let #app_var = #ty ::augment_clap(#app_var);
```
Comment (birkenfeld) [inline structopt-derive/src/lib.rs]: Note: this errors out with a somewhat strange message if `#ty` is something like `Option<Foo>`. Is there a good way to check for this?

**Comment by birkenfeld** [comment]
I'll add documentation and an example if you're okay with the idea and the implementation.

File: structopt-derive/src/lib.rs [hunk 2]
```diff
@@ -141,6 +151,8 @@ fn gen_constructor(fields: &Punctuated<Field, Comma>) -> quote::Tokens {
                 _ => quote!( .unwrap() )
             };
             quote!(#field_name: #subcmd_type::from_subcommand(matches.subcommand())#unwrapper)
+        } else if attrs.should_flatten() {
+            quote!(#field_name: StructOpt::from_clap(matches))
```
Comment (TeXitoi) [inline structopt-derive/src/lib.rs]: The usage seems always to be `if attrs.is_subcommand() { ... } else if attrs.should_flatten() { ... } else { ... }`.

Maybe an `enum Mode { SubCommand, Flatten, Arg }` (or other naming if you're more inspired than me) would be better?

Also, checking that we don't `subcommand` and `flatten` on the same field seems reasonable.

File: structopt-derive/src/lib.rs (hunk 1 above)
Comment (TeXitoi) [reply structopt-derive/src/lib.rs]:
```rust
let #app_var = <#ty>::augment_clap(#app_var);
```

Maybe you can update other similar code gen.

File: structopt-derive/src/attrs.rs [hunk 3]
```diff
@@ -150,7 +154,18 @@ impl Attrs {
...
+                    if let Kind::FlattenStruct = self.kind {
+                        panic!("subcommands cannot be flattened");
+                    }
+                    self.kind = Kind::Subcommand(Ty::Other);
+                }
+                Word(ref w) if w == "flatten" => {
+                    if let Kind::Subcommand(_) = self.kind {
+                        panic!("subcommands cannot be flattened");
+                    }
+                    self.kind = Kind::FlattenStruct;
```
Comment (TeXitoi) [inline structopt-derive/src/attrs.rs]: A setter with check can mutualize this code.

File: structopt-derive/src/lib.rs [hunk 4]
```diff
@@ -60,21 +60,24 @@ fn gen_augmentation(fields: &Punctuated<Field, Comma>, app_var: &Ident) -> quote
...
             } else {
                 quote!( let #app_var = #app_var.setting(::structopt::clap::AppSettings::SubcommandRequiredElseHelp); )
             };
 
             Some(quote!{
-                let #app_var = #subcmd_type ::augment_clap( #app_var );
+                let #app_var = <#subcmd_type>::augment_clap( #app_var );
                 #required
             })
+            } else {
```
Comment (TeXitoi) [inline structopt-derive/src/lib.rs]: Strange indentation.

**Comment by TeXitoi** [comment]
No more remarks after reading on a computer.

**Comment by TeXitoi** [comment]
v0.2.7 published

**Pull request merged**